x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java

```diff
@@ -388,45 +372,39 @@ protected Query asQuery(In in, TranslatorHandler handler) {
         }
 
         public static Query doTranslate(In in, TranslatorHandler handler) {
-            return handler.wrapFunctionQuery(in, in.value(), () -> translate(in, handler));
-        }
-
-        private static Query translate(In in, TranslatorHandler handler) {
-            Query q;
-            if (in.value() instanceof FieldAttribute) {
-                // equality should always be against an exact match (which is important for strings)
-                FieldAttribute fa = (FieldAttribute) in.value();
-                DataType dt = fa.dataType();
-
-                List<Expression> list = in.list();
-                Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-                list.forEach(e -> {
-                    // TODO: this needs to be handled inside the optimizer
-                    if (DataTypes.isNull(e.dataType()) == false) {
-                        set.add(handler.convert(valueOf(e), dt));
```

Comment:
I guess the [removed tests](https://github.com/elastic/elasticsearch/pull/76424/files#r688298282) were "enabled" by this - now removed - conversion, but I wonder why was it needed before: just for DateTime's? I guess the rest of the types are conveniently converted by ES.

Reply:
I think the way the implementation worked before assumed that all values in the list have the same type. But you can also mix datetimes and strings: `foo IN ('2020-10-01T01:00', '2020-10-01T20:00'::datetime)`. So the conversion was needed to enable these mixed cases with the particular implementation.

Reply:
> I think the way the implementation worked before assumed that all values in the list have the same type.

Not sure about that, from these existing test: [here](https://github.com/elastic/elasticsearch/blob/a5a1c6aa9148cd98a0cf3b139619e5fe7bd85dc2/x-pack/plugin/sql/qa/server/src/main/resources/filter.csv-spec#L134) and below.

In any case, this non-DateTime "mixed values" IN would remain supported after the PR, right? That should work as the values end up as strings in the query - with no further conversions/pre-processing - passed to ES, which does then itself the conversion to the field type. But DateTimes could be different, with maybe TZ adjustment.

Reply:
If this is an issue, it should be handled in a different PR (with a separate GH issue created).